lib/Sema/TypeCheckProtocol.cpp

```diff
@@ -5506,9 +5506,21 @@ ValueWitnessRequest::evaluate(Evaluator &eval,
   return known->second;
 }
 
-ValueDecl *TypeChecker::deriveProtocolRequirement(DeclContext *DC,
-                                                  NominalTypeDecl *TypeDecl,
-                                                  ValueDecl *Requirement) {
+ValueDecl *TypeChecker::deriveProtocolRequirement(
+    DeclContext *DC, NominalTypeDecl *TypeDecl, ValueDecl *Requirement,
+    NormalProtocolConformance *conformance) {
+  auto evaluator = DC->getASTContext().evaluator;
+  auto request = DeriveProtocolRequirementRequest{DC, TypeDecl, Requirement};
+  if (evaluator.hasActiveRequest(request) &&
+      !conformance->hasWitness(Requirement)) {
```

Comment:
Should an ongoing request not be enough to guarantee `!hasWitness`?

Reply:
I was looking at the logic for `isRecursiveValidation` but I suppose an active request should be enough.

Reply:
Can you add a FIXME comment to remove this check and break the cycle in a more principled manner?